langpuppet/gapi.go

```diff
@@ -0,0 +1,284 @@
+
+	for _, flag := range c.FlagNames() {
+		if !c.IsSet(flag) {
+			continue
+		}
+		childFlagName := strings.TrimPrefix(flag, FlagPrefix)
+		flagSet.String(childFlagName, "", "no usage string needed here")
+		flagSet.Set(childFlagName, c.String(flag))
+	}
+
+	var langDeploy *gapi.Deploy
+	var puppetDeploy *gapi.Deploy
+	var err error
+
+	// we don't really need the deploy object from the child GAPIs
+	if langDeploy, err = (&lang.GAPI{}).Cli(cli.NewContext(c.App, flagSet, nil), fs); err != nil {
+		return nil, err
+	}
+	if puppetDeploy, err = (&puppet.GAPI{}).Cli(cli.NewContext(c.App, flagSet, nil), fs); err != nil {
+		return nil, err
+	}
+
+	return &gapi.Deploy{
+		Name: Name,
+		Noop: c.GlobalBool("noop"),
+		Sema: c.GlobalInt("sema"),
+		GAPI: &GAPI{
+			InputURI:   fs.URI(),
+			langGAPI:   langDeploy.GAPI,
+			puppetGAPI: puppetDeploy.GAPI,
+		},
+	}, nil
+}
+
+// CliFlags returns a list of flags used by this deploy subcommand.
+// It consists of all flags accepted by lang and puppet mode,
+// with a respective "lp-" prefix.
+func (obj *GAPI) CliFlags() []cli.Flag {
+	langFlags := (&lang.GAPI{}).CliFlags()
+	puppetFlags := (&puppet.GAPI{}).CliFlags()
+
+	var childFlags []cli.Flag
+
+	for _, flag := range append(langFlags, puppetFlags...) {
+		childFlags = append(childFlags, &cli.StringFlag{
+			Name:  FlagPrefix + strings.Split(flag.GetName(), ",")[0],
+			Value: "",
+			Usage: fmt.Sprintf("equivalent for '%s' when using the lang/puppet entrypoint", flag.GetName()),
+		})
+	}
+
+	return childFlags
+}
+
+// Init initializes the langpuppet GAPI struct.
+func (obj *GAPI) Init(data gapi.Data) error {
+	if obj.initialized {
+		return fmt.Errorf("already initialized")
+	}
+	obj.data = data // store for later
+
+	dataLang := gapi.Data{
+		Program:       obj.data.Program,
+		Hostname:      obj.data.Hostname,
+		World:         obj.data.World,
+		Noop:          obj.data.Noop,
+		NoConfigWatch: obj.data.NoConfigWatch,
+		NoStreamWatch: obj.data.NoStreamWatch,
+		Debug:         obj.data.Debug,
+		Logf: func(format string, v ...interface{}) {
+			obj.data.Logf(lang.Name+": "+format, v...)
+		},
+	}
+
+	dataPuppet := gapi.Data{
+		Program:       obj.data.Program,
+		Hostname:      obj.data.Hostname,
+		World:         obj.data.World,
+		Noop:          obj.data.Noop,
+		NoConfigWatch: obj.data.NoConfigWatch,
+		NoStreamWatch: obj.data.NoStreamWatch,
+		Debug:         obj.data.Debug,
+		Logf: func(format string, v ...interface{}) {
+			obj.data.Logf(puppet.Name+": "+format, v...)
+		},
+	}
+
+	if err := obj.langGAPI.Init(dataLang); err != nil {
+		return err
+	}
+	if err := obj.puppetGAPI.Init(dataPuppet); err != nil {
+		return err
+	}
+
+	obj.closeChan = make(chan struct{})
+	obj.initialized = true
+	return nil
+}
+
+// Graph returns a current Graph.
+func (obj *GAPI) Graph() (*pgraph.Graph, error) {
+	if !obj.initialized {
+		return nil, fmt.Errorf("%s: GAPI is not initialized", Name)
+	}
+
+	var graph_err error
+
+	if obj.currentLangGraph == nil {
```

Comment:
I think the logic here should actually be something like this:

In Next(), whenever the corresponding lang or puppet child Next() call sends an event, set a "dirty" flag for that frontend. In the Graph() you run the Graph() function when that flag is true, and then unset it. (You might want to consider a lock.)

And you'll never send a message on Next() until you've gotten an initial message from each Next, and then you send on each subsequent message.

So this is a one line change here, and in Next for each. Let me know if this isn't clear enough.

Reply:
Done.